src/DataStructures/ComplexModalVector.hpp

```diff
@@ -0,0 +1,134 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+/// \file
+/// Defines class Data.
+
+#pragma once
+
+#include <complex>
+
+#include "DataStructures/ModalVector.hpp"
+#include "DataStructures/VectorImpl.hpp"
+#include "Utilities/PointerVector.hpp"  // IWYU pragma: keep
+#include "Utilities/TMPL.hpp"
+
+class ComplexDataVector;
+
+// IWYU pragma: no_include <blaze/math/expressions/DVecMapExpr.h>
+// IWYU pragma: no_include <blaze/math/typetraits/IsVector.h>
+// IWYU pragma: no_include "DataStructures/DataVector.hpp"
+// IWYU pragma: no_forward_declare ConstantExpressions_detail::pow
+
+/*!
+ * \ingroup DataStructuresGroup
+ * \brief A class for storing spectral coefficients on a mesh.
+ *
+ * \details A ComplexModalVector holds an complex array of spectral
+ * coefficients, and can be either owning (the array is deleted when the
+ * ModalVector goes out of scope) or non-owning, meaning it just has a pointer
+ * to an array.
+ *
+ * Only basic mathematical operations are supported with ComplexModalVectors. In
+ * addition to addition, subtraction, multiplication, division, there are the
+ * following element-wise operations:
+ *
+ * - max
+ * - min
+ */
+class ComplexModalVector
+    : public VectorImpl<std::complex<double>, ComplexModalVector> {
+ public:
+  using VectorImpl<std::complex<double>, ComplexModalVector>::operator=;
+  using VectorImpl<std::complex<double>, ComplexModalVector>::VectorImpl;
+
+  MAKE_EXPRESSION_MATH_ASSIGN_PV(+=, ComplexModalVector)
+  MAKE_EXPRESSION_MATH_ASSIGN_PV(-=, ComplexModalVector)
+};
+
+namespace blaze {
+template <>
+struct IsVector<ComplexModalVector> : std::true_type {};
+template <>
+struct TransposeFlag<ComplexModalVector>
+    : BoolConstant<ComplexModalVector::transpose_flag> {};
+
+BLAZE_TRAIT_SPEC_BINTRAIT(ComplexModalVector, AddTrait);
+BLAZE_TRAIT_SPEC_BINTRAIT(ComplexModalVector, SubTrait);
+BLAZE_TRAIT_COMPATIBLE_BINTRAIT(ComplexModalVector, ComplexDataVector,
```

Comment:
I'm also not sure we want mixing between `ComplexModalVector` and `ComplexDataVector` for the same reasons as modal and datavector.

Reply:
I can certainly understand the desire there, but for derivatives it will be very useful to have a vector of things to multiply the coefficients by and have an element-wise multiplication.

Is it perhaps a better way to go about it to have a `multiply_elementwise` function which does it explicitly (probably with the blaze-y upcasts) so that there's less chance it can be done carelessly with `*`?

Reply:
Aren't the derivatives computed by matrix multiplication? That's how it's done for dG anyway :) If not, then I think having an explicit function would be fine to avoid the incorrect accidental usage

Reply:
They are not. If they were, we wouldn't have to horse around with all the libsharp transformations at all ;) (note also the matrices would have to be ~(l*(l + 1))^2 in size). I'll add a explicit function when I clean up this PR.

Reply:
Haha ah okay, well I'm sorry :P Those are grossly large matrices :| 

Okay, that sounds like a good plan :)